router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
```

Comment:
one problem with that... there's still another subtle race. After the `select` in `RoundTrip()` enters the `case <-ctx.Done()`, we could get a response from `roundTrip()` and send that on the `resultc` before the request is actually canceled. That would result in leaving that response and its backend conn open forever.

Gonna take a fresh look at this after lunch to see if something like the they have in [`httputil.ReverseProxy`](https://github.com/golang/go/blob/fa851b17a2fd2350043b3e56dce398a606692c23/src/net/http/httputil/reverseproxy.go#L130-L156) might actually simplify this.